src/ledger/LedgerStateSnapshot.h

```diff
@@ -158,12 +165,43 @@ class LedgerSnapshot : public NonMovableOrCopyable
         return mGetter->getAccount(header, tx, AccountID);
     }
     LedgerEntryWrapper load(LedgerKey const& key) const;
+};
+
+// A subclass of LedgerSnapshot that contains additional snapshots necessary for
+// transaction validation.
+class ExtendedLedgerSnapshot : public LedgerSnapshot
+{
+  public:
+    // Generate a snapshot from the most recently closed ledger. The resulting
+    // `ExtendedLedgerSnapshot` cannot be used for transaction application.
+    explicit ExtendedLedgerSnapshot(Application const& app);
+
+    // Generate a snapshot from a specific AbstractLedgerTxn. Set `forApply` to
+    // `true` to use this snapshot in transaction application.
+    ExtendedLedgerSnapshot(AbstractLedgerTxn& ltx, AppConnector const& app,
+                           bool forApply);
+
+    // Generate a snapshot from a specific `AbstractLedgerTxn`, copying
+    // additional snapshot data from `outer`. This is intended only for use with
+    // `executeWithMaybeInnerSnapshot`.
+    ExtendedLedgerSnapshot(AbstractLedgerTxn& ltx,
+                           ExtendedLedgerSnapshot const& outer);
+    virtual ~ExtendedLedgerSnapshot() = default;
 
     // Execute a function with a nested snapshot, if supported. This is needed
     // to support the replay of old buggy protocols (<8), see
     // `TransactionFrame::loadSourceAccount`
     void executeWithMaybeInnerSnapshot(
-        std::function<void(LedgerSnapshot const&)> f) const;
-};
+        std::function<void(ExtendedLedgerSnapshot const&)> f) const;
 
+    // Getters for the additional snapshots
+    Config const& getConfig() const;
+    SorobanNetworkConfig const& getSorobanNetworkConfig() const;
+    uint32_t getCurrentProtocolVersion() const;
+
+  private:
+    std::shared_ptr<const Config> const mConfig;
+    std::optional<const SorobanNetworkConfig> const mSorobanNetworkConfig;
```

Comment:
I feel a bit uneasy SorobanNetworkConfig lives here: LedgerSnapshot was meant to be simply a ledger state wrapper, that supports both LedgerTxn and BucketListSnapshot underneath. I was hoping to not cache anything at this layer to avoid ledger state inconsistencies and other footguns. The only exception is mLedgerHeader, which has to be modified for validation purposes. SorobanNetworkConfig really belongs to immutable bucketlist snapshots, since the struct is just a different representation of what lives in the bucketlist already. Could we move it there?